### PR TITLE
[active-standby] Add oscillation logic when there is no heartbeat on both sides

### PR DESCRIPTION
**Pull request by zjswhhh**
### Description of PR

Summary:
Fixes # (issue)
This PR is to cover an extreme edge case when none of the dualtors can receive ICMP heartbeat. The goal is to oscillate mux direction between two sides so if one side recovers, mux direction can have a chance to park on this side. 

**Note** that toggles in an unhealthy scenario like this will cause longer disruption than in a healthy scenario. 

sign-off: Jing Zhang zhangjing@microsoft.com 

### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Doc/Design
- [ ] Unit test

#### What is the motivation for this PR?
To avoid mux direction from being parked on one side when missing heartbeat. 

##### Work item tracking
- Microsoft ADO **(number only)**:
25367027

**TODO:** will submit a separate PR for option to disable oscillation or increase interval. 

#### How did you do it?
1. Start a 5-min timer on active side when heartbeat is missing;
2. When timer expires if still missing heartbeat && still active, toggle to standby.

#### How did you verify/test it?
Run the change on lab devices. Oscillation happened like expected (adjusted interval to 1min on lab device): 
```
Nov  7 01:09:43.225217 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:11:59.883266 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:14:49.997030 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:17:45.068746 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:19:57.332493 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:22:14.209417 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:24:30.938922 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:26:47.640098 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:29:04.379680 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:31:13.777881 TOR-B NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation

Nov  7 01:08:41.815651 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:10:59.213299 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:13:16.606989 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:16:12.141991 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:18:53.202499 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:21:09.851763 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:23:26.517250 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:25:43.215731 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:27:59.887524 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
Nov  7 01:30:10.159067 TOR-A NOTICE mux#linkmgrd: DbInterface.cpp:237 handlePostSwitchCause: Ethernet124: post last switch cause Timed_Oscillation
```

### CODE REVIEW

**Review by yxieca — APPROVED** [review]

**Review by lolyu — CHANGES_REQUESTED** [review]
Can we have UT for this?

File: src/common/MuxConfig.h [hunk 1]
```diff
@@ -274,6 +274,27 @@ class MuxConfig
...
+    /**
+    *@method setOscillationInterval_sec
+    *
+    *@brief setter for mux state oscillation interval
+    *
+    *@param oscillationInterval_sec (in) oscillation interval
+    *
+    *@return none
+    */
+    inline void setOscillationInterval_sec(uint32_t oscillationInterval_sec) {mOscillationTimeout_sec = oscillationInterval_sec;};
```
Comment (lolyu) [inline src/common/MuxConfig.h:295]: Can we have a hard lower limit for the oscillation?

Comment (zjswhhh) [reply src/common/MuxConfig.h]: Addressed.

**Review by lolyu — APPROVED** [review]
LGTM

File: src/link_manager/LinkManagerStateMachineActiveStandby.cpp [hunk 2]
```diff
@@ -54,12 +54,15 @@ ActiveStandbyStateMachine::ActiveStandbyStateMachine(
...
-    mWaitTimer(strand.context())
+    mWaitTimer(strand.context()),
+    mOscillationTimer(strand.context())
 {
     assert(muxPortPtr != nullptr);
     mMuxStateMachine.setWaitStateCause(mux_state::WaitState::WaitStateCause::SwssUpdate);
     mMuxPortPtr->setMuxLinkmgrState(mLabel);
     initializeTransitionFunctionTable();
+
+    mOscillationTimer.expires_from_now(boost::posix_time::seconds(1));
```
Comment (lolyu) [inline src/link_manager/LinkManagerStateMachineActiveStandby.cpp:65]: why do we need to set the timer expiry time here?

Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachineActiveStandby.cpp:65]: > why do we need to set the timer expiry time here?

To initiate the expiry time, later in Line 1304 we compare expiry time to current local time before starting the timer.

**Pull request merged**